### ISSUE TITLE
Hot vs cold observable

### ISSUE THREAD

**Issue by teivah** [ReactiveX/RxGo#179]
In the v2 branch we started to make a clear distinction between hot and cold observable. This issue is a placeholder to discuss about the ongoing implementation, the API, the impacts etc.

**Comment by teivah** [comment]
We started to make the distinction between cold and hot observable:
* Cold: created with cold operators like `Just` for example. Allows an observer to subscribe at any time and to be able to see the whole history. The items are pulled when the observer subscribes to an observable.
* Hot: created with hot operator like `FromEventSource`. The items are pushed regardless of the subscriptions. When an observer subscribes, it will not see the items produced in the past.

See the current tests: https://github.com/ReactiveX/RxGo/blob/6c1de57ec0e14f0cbccc951ed5a6b200dd211ef2/connectableobservable_test.go#L9-L70

Any feedback is welcome.

**Comment by chgeuer** [comment]
I started to play with v2 and noticed that many of the operators are hard-coded for cold observables. For example, looking at [`Filter`](https://github.com/ReactiveX/RxGo/blob/v2/observable.go#L825-L840), it always returns a `newColdObservableFromFunction()`. In my code, I actually need to support hot observables, and changed the `Filter()` implementation like so: 

```go
// Filter filters items in the original Observable and returns
// a new Observable with the filtered items.
func (o *observable) Filter(apply Predicate) Observable {
	if o.observableType == cold {
		//
		// The previous implementation here 
		// 
		...

		return newColdObservableFromFunction(f)
	}

	out := make(chan interface{})
	o.Subscribe(handlers.NextFunc(func(item interface{}) {
		if apply(item) {
			out <- item
		}
	}))
	return newHotObservableFromChannel(out)
}
```

That approach also seems to work for example for [`Map()`](https://github.com/chgeuer/go_hotfolder_azure/blob/f528cf219db428b2253bdd37409b326c4983a1af/vendor/github.com/reactivex/rxgo/observable.go#L947-L968): 

```go
func (o *observable) Map(apply Function) Observable {
	if o.observableType == cold {
		// Old impl here
		...
		return newColdObservableFromFunction(f)

	}

	out := make(chan interface{})
	o.Subscribe(handlers.NextFunc(func(item interface{}) {
		out <- apply(item)
	}))
	return newHotObservableFromChannel(out)
}
```

My question is: Is that the proper RxGo way, or do you foresee a different implementation strategy to deal with cold and hot observables?

**Comment by jochasinga** [comment]
When I first started building RxGo, I got my inspiration from data flow programming like Max/MSP and Pure Data. I encourage anyone to read up on it. At a conceptual level, data flow programming is very close to reactive stream.

**Comment by teivah** [comment]
> My question is: Is that the proper RxGo way, or do you foresee a different implementation strategy to deal with cold and hot observables?

This was a first draft I would say, the issue itself is still opened and every feedback like yours is more than welcome. Anyway, you are right there are still work to do to support your use case. Maybe with if/else statements or by having two different observable interfaces...

**Comment by cerlestes** [comment]
My two cents: I really liked the decision.

With my current hobby project I started working on a minimal implementation of the actor model in golang, just for fun, and I wanted to build it around the reactive extensions family of libraries, because I want to get familiar with its API.

If I'm honest, I had quite a bit of trouble getting into v1 of RxGo, and I already wanted to divert away from the project and also aim for my own quick and minimal implementation as I needed it. The major problem was that from my tests, an Observable seemed to publish incoming events to only one subscriber, because fundamentally it was just a channel and behaved that way. The other problem I had was with the structure of the API being very complicated.

A bit frustrated I was delighted to find the v2 branch and started looking into it. And I have to say: the current v2 version works beautifully for my project. The notion of hot vs cold observable is important, especially in go with its abundance of short lived channels.

And thanks to especially `rxgo.FromEventSource()`, I was able to simplify my code regarding RxGo a lot (like ~20 lines into just that call). The shallow module structure and resulting easy API makes working with RxGo as fun as it deserves to be. And everything now is working as expected. So hats off to the great rewrite and thank you a lot for your continuous work! :heart:
